### PR TITLE
Forward defined types not properly defined in TypeScript - Close 633

### PR DESCRIPTION
**Pull request by pepone**
This pull request fixed issue described in #633

### CODE REVIEW

File: cpp/src/slice2js/Gen.cpp [hunk 1]
```diff
@@ -2293,12 +2293,31 @@ Slice::Gen::TypeScriptRequireVisitor::addImport(const TypePtr& definition, const
...
+                if(!definedIn.empty())
+                {
+                    p1 = definedIn;
+                }
+            }
+            else
+            {
+                m1 = getModuleMetadata(ContainedPtr::dynamicCast(c->definition()));
+                p1 = c->definition()->definitionContext()->filename();
+            }
```
Comment (bernardnormier) [inline cpp/src/slice2js/Gen.cpp]: Do we need this else? When in the else, the class/interface is fully defined, which should be just like before this patch.

File: cpp/src/slice2js/JsUtil.cpp [hunk 2]
```diff
@@ -164,6 +164,24 @@ fixIds(const StringList& ids)
...
+string
+Slice::JsGenerator::getDefinedIn(const ContainedPtr& p)
+{
+    static const string prefix = "js:defined-in:";
+    string definedIn;
+    if(findMetaData(prefix, p->getMetaData(), definedIn))
+    {
+        string base = p->definitionContext()->filename();
+        string::size_type pos = base.find_last_of("/\\");
+        if(pos != string::npos)
```
Comment (bernardnormier) [inline cpp/src/slice2js/JsUtil.cpp]: Should be add an assert if no / or \ is found? In this case, the resulting string would not formed correctly.

File: cpp/src/slice2js/Gen.cpp (hunk 1 above)
Comment (pepone) [reply cpp/src/slice2js/Gen.cpp]: Without this for a forward declared type that is already defined, we import the file corresponding to the forward declaration and not the one corresponding to the definition. It was bogus before.

File: cpp/src/slice2js/JsUtil.cpp (hunk 2 above)
Comment (pepone) [reply cpp/src/slice2js/JsUtil.cpp]: I think we can add the assert, the filename is always a full-path if I'm not mistaken, I will check.

**Pull request merged**